src/psyclone/tests/domain/nemo/transformations/acc_update_test.py

```diff
@@ -0,0 +1,463 @@
+            "  checksum = sum(zftv)\n"
+            "  !$acc update if_present device(checksum)\n" in gen_code)
+
+
+def test_loop_on_cpu(parser):
+    ''' Test the application of ACCUpdateTrans to CPU code containing a
+    loop which itself contains a Call. '''
+    code = '''
+SUBROUTINE tra_ldf_iso()
+  USE some_mod, only: dia_ptr_hst
+  INTEGER, PARAMETER :: jpi=2, jpj=2, jpk=2, start=2, step=2
+  INTEGER :: jn, ji
+  LOGICAL :: l_ptr
+  INTEGER, DIMENSION(jpi,jpj) :: tmask
+  REAL, DIMENSION(jpi,jpj,jpk) :: zdit, zdjt, zftu, zftv, zftw
+  zftv(:,:,:) = 0.0d0
+  DO ji = start, jpi, step
+    CALL dia_ptr_hst( ji, 'ldf', zftv(ji,:,:)  )
+    zftv(ji,:,:) = 1.0d0
+    zftw(ji,:,:) = -1.0d0
+  END DO
+  zftu(:,:,1) = 1.0d0
+  tmask(:,:) = jpi
+end SUBROUTINE tra_ldf_iso
+'''
+    reader = FortranStringReader(code)
+    ast = parser(reader)
+    psy = PSyFactory(API, distributed_memory=False).create(ast)
+    schedule = psy.invokes.invoke_list[0].schedule
+    acc_update = ACCUpdateTrans()
+    acc_update.apply(schedule)
+    gen_code = str(psy.gen).lower()
+    # Loop variable should not be copied to device
+    assert "device(ji)" not in gen_code
+    # TODO Currently jpi is copied to and from. We need a list of variables
+    # that are written just once and are read-only thereafter. Such variables
+    # will be written on the CPU (e.g. after reading from namelist).
+    assert ("  !$acc update if_present host(jpi,start,step,zftv)\n"
+            "  zftv(:,:,:) = 0.0d0\n"
+            "  !$acc update if_present device(zftv)\n"
+            "  do ji = start, jpi, step\n") in gen_code
+    assert ("    !$acc update if_present host(zftv)\n"
+            "    call") in gen_code
+    assert ("    !$acc update if_present host(zftv,zftw)\n"
+            "    zftv(ji,:,:) = 1.0d0\n"
+            "    zftw(ji,:,:) = -1.0d0\n"
+            "    !$acc update if_present device(zftv,zftw)\n") in gen_code
+    # TODO All of these variables are actually local to the subroutine so
+    # should not be copied back.
+    assert ("  !$acc update if_present host(jpi,tmask,zftu)\n"
+            "  zftu(:,:,1) = 1.0d0\n"
+            "  tmask(:,:) = jpi\n"
+            "  !$acc update if_present device(tmask,zftu)\n"
+            in gen_code)
+
+
+def test_codeblock(parser):
+    ''' Test that we can reason about CodeBlocks.'''
+    code = '''
+subroutine lbc_update()
+  use oce, only: tmask, jpi, jpj, jpk
+
+  open(unit=32, file="some_forcing.dat")
+  read(32, *) jpi, jpj, jpk
+  allocate(tmask(jpi, jpj))
+  read(32, *) tmask
+
+end subroutine lbc_update
+'''
+    reader = FortranStringReader(code)
+    ast = parser(reader)
+    psy = PSyFactory(API, distributed_memory=False).create(ast)
+    schedule = psy.invokes.invoke_list[0].schedule
+    acc_update = ACCUpdateTrans()
+    acc_update.apply(schedule)
+    assert isinstance(schedule[1], CodeBlock)
+    gen_code = str(psy.gen).lower()
+    # TODO We are probably relying on undefined behaviour here
```

Comment:
No, it was because, when I wrote that, I thought `Codeblock.get_symbol_names()` would also retrieve other symbols beyond variable names, like Fortran statements or called intrinsics, functions, subroutines. At the moment, we do incur the risk of collecting a function name mistaken by an array name, but that's already documented in the code, i.e. in `acc_update_trans.py`, including a reference to #1872. I've removed it as this codeblock calls no user-defined functions, rendering the comment irrelevant.